src/ophyd_async/core/_device.py

```diff
@@ -1,128 +1,155 @@
-    _previous_connect_was_mock = None
+    parent: DeviceBase | None = None
 
-    def __init__(self, name: str = "") -> None:
+    def __init__(
+        self,
+        name: str = "",
+    ) -> None:
+        self._connect_cache = DeviceConnectCache()
         self.set_name(name)
 
     @property
     def name(self) -> str:
         """Return the name of the Device"""
         return self._name
 
-    @cached_property
-    def log(self):
-        return LoggerAdapter(
+    def set_name(self, name: str):
+        self._name = name
+        # Ensure self.log is recreated after a name change
+        self.log = LoggerAdapter(
             getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
         )
 
-    def children(self) -> Iterator[tuple[str, "Device"]]:
-        for attr_name, attr in self.__dict__.items():
-            if attr_name != "parent" and isinstance(attr, Device):
-                yield attr_name, attr
 
-    def set_name(self, name: str):
-        """Set ``self.name=name`` and each ``self.child.name=name+"-child"``.
+DeviceBaseT = TypeVar("DeviceBaseT", bound=DeviceBase)
 
-        Parameters
-        ----------
-        name:
-            New name to set
-        """
 
-        # Ensure self.log is recreated after a name change
-        if hasattr(self, "log"):
-            del self.log
+class DeviceBackend:
+    def __init__(self, device_type: type[Device]):
+        self.device_type = device_type
+        self.children: dict[str, DeviceBase] = {}
 
-        self._name = name
-        for attr_name, child in self.children():
-            child_name = f"{name}-{attr_name.rstrip('_')}" if name else ""
-            child.set_name(child_name)
-            child.parent = self
+    # TODO: we will add some mechanism of invalidating the cache here later
+    async def connect(self, mock: bool, timeout: float, force_reconnect: bool) -> None:
+        coros = {
+            name: child_device.connect(
+                mock=mock, timeout=timeout, force_reconnect=force_reconnect
+            )
+            for name, child_device in self.children.items()
+        }
+        await wait_for_connection(**coros)
+
+
+class Device(DeviceBase):
+    """Common base class for all Ophyd Async Devices."""
+
+    _backend: DeviceBackend
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls)
+        # Need to make the backend first in case any children are
+        # created before the super().__init__ call
+        instance._backend = DeviceBackend(cls)  # noqa: SLF001
+        return instance
+
+    def __init__(self, name: str = "", backend: DeviceBackend | None = None) -> None:
+        if backend:
+            # Copy into the new backend the children that have already been added
+            backend.children.update(self._backend.children)
+            self._backend = backend
+        super().__init__(name)
+
+    def _set_child_name(self, child: DeviceBase, child_name: str):
```

Comment:
Are the annotations in here `DeviceBase` instead of `Device` because we expect others to implement their own `Device`?